perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -518,6 +643,240 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
                   host_to_connect=host_to_connect)
 
 
+def RunThreadedWorkers(worker, worker_args, per_thread_args=None):
+  """Run a worker function in many threads, then gather and return the results.
+
+  Args:
+    worker: either WriteWorker or ReadWorker. The worker function to call.
+    worker_args: a tuple. The arguments to pass to the worker function. The
+      result queue and stream number will be appended as the last two arguments.
+    per_thread_args: if given, an array with length equal to the
+      number of threads. Thread number i will be passed
+      per_thread_args[i] after its regular arguments and before the
+      result queue and stream number.
+
+  Returns:
+    A list of the results returned by the workers.
+  """
+
+  result_queue = Queue.Queue()
+
+  logging.info('Creating %s threads', FLAGS.num_streams)
+  if per_thread_args is None:
+    threads = [threading.Thread(target=worker,
+                                args=worker_args + (result_queue, i))
+               for i in xrange(FLAGS.num_streams)]
+  else:
+    threads = [threading.Thread(target=worker,
+                                args=worker_args +
+                                (per_thread_args[i],) +
+                                (result_queue, i))
+               for i in xrange(FLAGS.num_streams)]
+  logging.info('Threads created. Starting threads.')
+  for thread in threads:
+    thread.start()
+  logging.info('Threads started.')
+  while True:
+    time.sleep(THREAD_STATUS_LOG_INTERVAL)
+    num_alive = sum((thread.isAlive() for thread in threads))
+    if num_alive == 0:
+      break
+    else:
+      logging.info('%s of %s threads are still working.',
+                   num_alive, FLAGS.num_streams)
+  logging.info('All threads complete.')
+
+  results = []
+
+  try:
+    while True:
+      res = result_queue.get_nowait()
+      results.append(res)
+  except Queue.Empty:
+    pass
+
+  return results
+
+
+def MultiStreamWrites(storage_schema, host_to_connect=None):
+  """Run multi-stream write threads."""
+
+  size_distribution = yaml.load(cStringIO.StringIO(FLAGS.object_sizes))
+
+  payload_bytes, payload_string = GenerateWritePayload(
+      MaxSizeInDistribution(size_distribution))
+
+  results = RunThreadedWorkers(
+      WriteWorker,
+      (storage_schema,
```

Comment:
where and when do you pass in the bucket?! (which is expected by the WriteWorker?)

Have you tried to run this code at all?